cclib/method/bader.py

```diff
@@ -0,0 +1,223 @@
+
+                            elif (
+                                xindex + nextDirection[0] == 0
+                                or xindex + nextDirection[0] == (len(self.chgdensity.data) - 1)
+                                or yindex + nextDirection[1] == 0
+                                or yindex + nextDirection[1] == (len(self.chgdensity.data[0]) - 1)
+                                or zindex + nextDirection[2] == 0
+                                or zindex + nextDirection[2]
+                                == (len(self.chgdensity.data[0][0]) - 1)
+                            ):
+                                # Avoid exiting the grid
+                                baderIndex = nextIndex
+                                nexIndex += 1
+
+                                listcoord = tuple(np.array(listcoord).T)
+                                self.fragresults[listcoord] = baderIndex
+
+                                loopcondition = False
+
+                            else:
+                                # Advance to the next point according to the direction of
+                                # maximum gradient
+                                xindex += nextDirection[0]
+                                yindex += nextDirection[1]
+                                zindex += nextDirection[2]
+
+        # Now try to identify each Bader region to individual atom.
+        # Try to find an area that captures enough representation
+        self.matches = numpy.zeros_like(self.data.atomnos)
+        for pos in range(len(self.data.atomcoords[-1])):
+            gridpt = numpy.round(
+                (self.data.atomcoords[-1][pos] - self.volume.origin) / self.volume.spacing
+            )
+
+            xgrid = int(gridpt[0])
+            ygrid = int(gridpt[1])
+            zgrid = int(gridpt[2])
+
+            self.matches[pos] = self.fragresults[xgrid, ygrid, zgrid]
+
+        assert (
+            0 not in self.matches
+        ), "Failed to assign Bader regions to atoms. Try with a finer grid. Content of Bader area matches: {}".format(
+            matches
+        )
+        assert len(
+            numpy.unique(self.matches) != len(self.data.atomnos)
+        ), "Failed to assign unique Bader regions to each atom. Try with a finer grid."
+
+        # Finally integrate the assigned Bader areas
+        self.logger.info("Creating fragcharges: array[1]")
+        boxvol = (
+            self.volume.spacing[0]
+            * self.volume.spacing[1]
+            * self.volume.spacing[2]
+            * convertor(1, "Angstrom", "bohr") ** 3
+        )
+        self.fragcharges = numpy.zeros(len(self.data.atomcoords[-1]), "d")
+
+        for atomIndex, baderareaIndex in enumerate(self.matches):
+            # turn off all other grid points
+            vals = numpy.copy(self.chgdensity.data)
+            mask = self.fragresults == baderareaIndex
+            vals *= mask
+            self.fragcharges[atomIndex] = sum(vals.ravel()) * boxvol
```

Comment:
Is there a way to rewrite this using `Volume.integrate()`? (I can't tell, kinda tired...)

Reply:
It is definitely possible and I've considered using `Volume.integrate()`. I think to use `Volume.integrate()`, though, n (# of atoms) number of Volume objects would have to be created each with appropriate masks to call on the method. So line 218-220 would still be there and line 221 would only be modified to a constructor for `Volume` object and call to `integrate()` method.

I think using `integrate()` method here would make it easier to follow any future changes in the method/object at an expense of a few more lines of code. I can make the necessary modifications if this fits better. Thanks!

Reply:
Yes, try it and see what it looks like.

Reply:
I've modified this part of the code to use `integrate()` method. Let me know if this is good to go!